File: scripts/python/category_utils.py
#!/usr/bin/env python3
"""
Category Utilities

Utilities for working with the centralized category configuration.
This is used by both the JavaScript and Python scripts to ensure consistency.
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Union

# Get the root directory of the project
def get_project_root() -> Path:
    """Get the root directory of the project."""
    current_dir = Path(os.path.dirname(os.path.abspath(__file__)))
    return current_dir.parent.parent

def load_categories_config() -> Dict:
    """
    Load the categories from the central configuration file
    
    Returns:
        Dict: The categories configuration
    """
    config_path = get_project_root() / 'config' / 'categories.json'
    
    try:
        if not config_path.exists():
            print(f"Error: Categories configuration file not found at {config_path}")
            return create_default_config()
        
        with open(config_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading categories configuration: {str(e)}")
        return create_default_config()

def create_default_config() -> Dict:
    """
    Create a default configuration if the config file doesn't exist
    
    Returns:
        Dict: Default configuration
    """
    return {
        "version": "1.0",
        "lastUpdated": datetime.now().isoformat(),
        "categories": [
            {
                "id": "ai",
                "name": "Artificial Intelligence",
                "keywords": ["machine learning", "AI"],
                "color": "#3498db"
            },
            {
                "id": "programming",
                "name": "Programming",
                "keywords": ["javascript", "python"],
                "color": "#2ecc71"
            }
        ],
        "searchSuffixes": ["filetype:pdf"]
    }

def update_categories_config(config: Dict) -> bool:
    """
    Update the categories configuration file
    
    Args:
        config: The new configuration to save
        
    Returns:
        bool: Whether the update was successful
    """
    config_path = get_project_root() / 'config' / 'categories.json'
    
    try:
        # Ensure directory exists
        os.makedirs(os.path.dirname(config_path), exist_ok=True)
        
        # Update lastUpdated timestamp
        config["lastUpdated"] = datetime.now().isoformat()
        
        # Write the file
        with open(config_path, 'w', encoding='utf-8') as f:
            json.dump(config, f, indent=2)
        
        return True
    except Exception as e:
        print(f"Error updating categories configuration: {str(e)}")
        return False

def detect_categories(text: str) -> List[str]:
    """
    Detect category IDs based on text content
    
    Args:
        text: Text to analyze for category keywords
        
    Returns:
        List[str]: Array of category IDs matching the text
    """
    config = load_categories_config()
    text_lower = text.lower()
    matches = []
    
    for category in config["categories"]:
        for keyword in category["keywords"]:
            if keyword.lower() in text_lower:
                if category["id"] not in matches:
                    matches.append(category["id"])
                break
    
    return matches

def generate_search_queries(category_id: str) -> List[str]:
    """
    Generate search queries for a category based on the configuration
    
    Args:
        category_id: The category ID to generate queries for
        
    Returns:
        List[str]: Array of search queries
    """
    config = load_categories_config()
    category = next((cat for cat in config["categories"] if cat["id"] == category_id), None)
    
    if not category:
        return []
    
    # Generate search queries by combining keywords with suffixes
    queries = []
    
    for keyword in category["keywords"]:
        for suffix in config["searchSuffixes"]:
            queries.append(f"{keyword} {suffix}")
    
    return queries

def get_category_by_id(category_id: str) -> Optional[Dict]:
    """
    Get a category by its ID
    
    Args:
        category_id: The category ID to look for
        
    Returns:
        Optional[Dict]: The category object or None if not found
    """
    config = load_categories_config()
    return next((cat for cat in config["categories"] if cat["id"] == category_id), None)

def ensure_json_schema_compatibility(pdf_entry: Dict) -> Dict:
    """
    Ensure a PDF entry conforms to the JavaScript schema
    
    Args:
        pdf_entry: The PDF entry to normalize
        
    Returns:
        Dict: The normalized PDF entry
    """
    # Required fields with default values
    schema_fields = {
        "id": None,  # Should be provided
        "url": None,  # Should be provided
        "title": "Untitled PDF",
        "author": "Unknown",
        "categories": [],
        "source": "unknown",
        "yearPublished": None,
        "tags": [],
        "isAvailable": True,
        "dateAdded": datetime.now().strftime("%Y-%m-%d"),
        "lastChecked": datetime.now().strftime("%Y-%m-%d"),
        "lastStatus": 200,
        "pages": None,
        "sizeMB": None
    }
    
    # Apply defaults for missing fields
    for field, default_value in schema_fields.items():
        if field not in pdf_entry:
            pdf_entry[field] = default_value
    
    # Ensure arrays are arrays
    for field in ["categories", "tags"]:
        if not isinstance(pdf_entry[field], list):
            if pdf_entry[field]:
                pdf_entry[field] = [pdf_entry[field]]
            else:
                pdf_entry[field] = []
    
    # Remove non-standard fields
    allowed_fields = set(schema_fields.keys())
    for field in list(pdf_entry.keys()):
        if field not in allowed_fields:
            del pdf_entry[field]
    
    return pdf_entry

# Direct command line usage to update search queries
if __name__ == "__main__":
    config = load_categories_config()
    
    print(f"Loaded {len(config['categories'])} categories from configuration")
    
    # If arguments are provided, generate search queries for that category
    if len(sys.argv) > 1:
        category_id = sys.argv[1]
        queries = generate_search_queries(category_id)
        
        print(f"\nSearch queries for category '{category_id}':")
        for query in queries:
            print(f"- {query}")
    else:
        # Otherwise, print all categories
        print("\nAvailable categories:")
        for category in config["categories"]:
            print(f"- {category['id']}: {category['name']} ({len(category['keywords'])} keywords)") 

File: scripts/js/category-utils.js
/**
 * Category Utilities
 * 
 * Utilities for working with the centralized category configuration.
 * This is used by both the JavaScript and Python scripts to ensure consistency.
 */

const fs = require('fs');
const path = require('path');

/**
 * Load the categories from the central configuration file
 * 
 * @returns {Object} The categories configuration
 */
function loadCategoriesConfig() {
  const configPath = path.resolve(__dirname, '../../config/categories.json');
  
  try {
    if (!fs.existsSync(configPath)) {
      console.error(`Error: Categories configuration file not found at ${configPath}`);
      return createDefaultConfig();
    }
    
    const configData = fs.readFileSync(configPath, 'utf8');
    return JSON.parse(configData);
  } catch (error) {
    console.error(`Error loading categories configuration: ${error.message}`);
    return createDefaultConfig();
  }
}

/**
 * Create a default configuration if the config file doesn't exist
 * 
 * @returns {Object} Default configuration
 */
function createDefaultConfig() {
  return {
    version: "1.0",
    lastUpdated: new Date().toISOString(),
    categories: [
      { id: "ai", name: "Artificial Intelligence", keywords: ["machine learning", "AI"], color: "#3498db" },
      { id: "programming", name: "Programming", keywords: ["javascript", "python"], color: "#2ecc71" }
    ],
    searchSuffixes: ["filetype:pdf"]
  };
}

/**
 * Update the categories configuration file
 * 
 * @param {Object} config The new configuration to save
 * @returns {boolean} Whether the update was successful
 */
function updateCategoriesConfig(config) {
  const configPath = path.resolve(__dirname, '../../config/categories.json');
  
  try {
    // Ensure directory exists
    const dirPath = path.dirname(configPath);
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }
    
    // Update lastUpdated timestamp
    config.lastUpdated = new Date().toISOString();
    
    // Write the file
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    return true;
  } catch (error) {
    console.error(`Error updating categories configuration: ${error.message}`);
    return false;
  }
}

/**
 * Detect category IDs based on text content
 * 
 * @param {string} text Text to analyze for category keywords
 * @returns {string[]} Array of category IDs matching the text
 */
function detectCategories(text) {
  const config = loadCategoriesConfig();
  const textLower = text.toLowerCase();
  const matches = [];
  
  for (const category of config.categories) {
    for (const keyword of category.keywords) {
      if (textLower.includes(keyword.toLowerCase())) {
        if (!matches.includes(category.id)) {
          matches.push(category.id);
        }
        break;
      }
    }
  }
  
  return matches;
}

/**
 * Generate search queries for a category based on the configuration
 * 
 * @param {string} categoryId The category ID to generate queries for
 * @returns {string[]} Array of search queries
 */
function generateSearchQueries(categoryId) {
  const config = loadCategoriesConfig();
  const category = config.categories.find(cat => cat.id === categoryId);
  
  if (!category) {
    return [];
  }
  
  // Generate search queries by combining keywords with suffixes
  const queries = [];
  
  for (const keyword of category.keywords) {
    for (const suffix of config.searchSuffixes) {
      queries.push(`${keyword} ${suffix}`);
    }
  }
  
  return queries;
}

/**
 * Get the category list in the format expected by the JavaScript crawler
 * 
 * @returns {Array} Array of category objects
 */
function getJavaScriptCategories() {
  const config = loadCategoriesConfig();
  return config.categories;
}

/**
 * Update the search-config.js file to use the central categories
 */
function updateSearchConfig() {
  const config = loadCategoriesConfig();
  const searchConfigPath = path.resolve(__dirname, './search-config.js');
  
  try {
    // Read the current search-config.js
    const currentConfig = fs.readFileSync(searchConfigPath, 'utf8');
    
    // Split out the categories section
    const categoriesStart = currentConfig.indexOf('categories: [');
    const categoriesEnd = currentConfig.indexOf('],', categoriesStart) + 1;
    
    // Generate the new categories section
    const categoriesSection = `categories: ${JSON.stringify(config.categories, null, 2)}`;
    
    // Replace the section in the file
    const updatedConfig = 
      currentConfig.substring(0, categoriesStart) + 
      categoriesSection + 
      currentConfig.substring(categoriesEnd);
    
    // Write the updated config
    fs.writeFileSync(searchConfigPath, updatedConfig);
    
    console.log('Updated search-config.js with centralized categories');
    return true;
  } catch (error) {
    console.error(`Error updating search-config.js: ${error.message}`);
    return false;
  }
}

/**
 * Update the searches.md file to match the centralized categories
 */
function updateSearchesFile() {
  const config = loadCategoriesConfig();
  const searchesFilePath = path.resolve(__dirname, '../../config/searches.md');
  
  try {
    // Generate the new content
    let content = '# PDF Search Queries\n\n';
    
    for (const category of config.categories) {
      content += `# ${category.name}\n`;
      
      // Generate 5 search queries for each category
      const queries = generateSearchQueries(category.id).slice(0, 5);
      for (const query of queries) {
        content += `${query}\n`;
      }
      
      content += '\n';
    }
    
    // Write the file
    fs.writeFileSync(searchesFilePath, content);
    
    console.log('Updated searches.md with centralized categories');
    return true;
  } catch (error) {
    console.error(`Error updating searches.md: ${error.message}`);
    return false;
  }
}

module.exports = {
  loadCategoriesConfig,
  updateCategoriesConfig,
  detectCategories,
  generateSearchQueries,
  getJavaScriptCategories,
  updateSearchConfig,
  updateSearchesFile
}; 